Source/WebKit/Scripts/webkit/messages.py

```diff
@@ -152,16 +157,31 @@ def function_parameter_type(type, kind):
     if type in builtin_types:
         return type
 
+    if type in types_that_must_be_moved():
```

Comment:
I think the IPC layer shouldn't enforce the "must be moved" on the count of:
 - Feature would be useful for copyable types
 - Theoretical perf difference since the caller can move the object into the call frame and doesn't need own storage

The type itself should enforce whether it's ok to copy.

So it would be better to frame it as:
- "Support for rvalue types" (as opposed to "Support move only types")
- is_by_value_type(name, kind) (as opposed to builtin_types, kind.startswith and types_that_must_be_moved)
- use std::forward if is_by_value_type() matches

Reply:
Sorry, I'm not sure I'm following what the code change would be for this.

Do we want to actually hold the object by value somewhere? I though the current approach of just forwarding the rval-reference through to the `encode` implementation was going to have the lowest overhead.

Reply:
Rvalue ref is a reference, e.g. a pointer to object that lives in somewhere.
When we hold by value, the object doesn't have to neccessarily live anywhere. 
So case like
```
send(Msg { funcReturningAHandle() })
-->
By value:
send(Msg { funcReturningAHandle() })
By reference:
auto h = funcReturningAHandle();
send(Msg { WTFMove(h) });
```

The former can operate on the value. If a move happens, it's a conceptual "no-op".
The later operates on the pointer. If move happens, it's a write to the object pointed via the pointer (the ref).

Reply:
Right, I think I'm mostly with you :)

Does that still work when `Msg` isn't an aggregate type (has a user-declared constructor), so we can't use `{ }` initialisation (I don't think?)?

If the constructor takes the object by value, and then moves the ctor parameter into `m_arguments`, would `funcReturningAHandle()` still avoid the temporary living somewhere?